ADR/0014-let-pipelines-proceed.md

```diff
@@ -0,0 +1,53 @@
+# 14. Let Pipelines Proceed
+
+* Date Decided: 2022-05-??
+* Date Documented: 2023-01-31
+
+## Status
+
+Accepted
+
+Relates to [ADR 13. Stonesoup Test Stream - API contracts](0013-integration-service-api-contracts.html)
+
+## Context
+
+The user's build pipeline includes scanning and linting tasks that operate on the source code and
+the built image (SAST, antivirus, clair, etc..). The purpose of these tasks is to find problems in
+the user's source code or dependencies and alert the user so they can take action ([STONE-459]).
+
+One frustration we've heard from users in previous systems is that they don't want to be blocked in
+their development or testing by complications in the build system, by compliance concerns. We want
+to fix that by offering a system that permits as much progress in the lifecycle of the user's
+application (build, test, and pre-production deployment) but which also protects production from
+non-compliant builds via mechanisms in the [enterprise contract].
+
+A problem we face: in Tekton, a failing TaskRun causes the whole PipelineRun to fail. If the purpose
+of our linting and scanning tasks is to find problems - which usually looks like failure with
+a non-zero exit code - how do they do their job without constantly breaking the user's build,
+stopping them from running integration tests, and stopping them from deploying candidate builds to
+their lower [Environments]?
+
+## Decision
+
+All scanning and linting TaskRuns should *succeed* even if they find problems in the content they
+are evaluating.
+
+Use the `HACBS_TEST_OUTPUT` result convention from [ADR-0013] to expose those results and render them
+for users ([STONE-459]).
```

Comment:
Would it be up to both the UI and PaC to render the output to users? I would expect that at least some GitHub users/organizations would want to enforce checks to pass before merging. How might this be able to be accomplished?

Instead of just having the UI/PaC go through all TaskRuns, would it be reasonable to have a "finalizer" Task which can aggregate and summarize the result of the previous scanning/linting TaskRuns. 

I don't _really_ like stating a further reliance on `HACBS_TEST_OUTPUT` as I think that we should try to come to some other generic pattern which doesn't propagate a tie-down to an old product name.

Reply:
> I don't really like stating a further reliance on HACBS_TEST_OUTPUT as I think that we should try to come to some other generic pattern which doesn't propagate a tie-down to an old product name.

Agreed. It is one of the reasons we've been holding off on moving forwards with [STONE-459](https://issues.redhat.com/browse/STONE-459), but it's getting too late now and I don't want to wait for [TKNECO-12](https://issues.redhat.com/browse/TKNECO-12). Once we get a recommendation there, we should land new decisions overturning this ADR and ADR 13.

> Would it be up to both the UI and PaC to render the output to users? I would expect that at least some GitHub users/organizations would want to enforce checks to pass before merging. How might this be able to be accomplished?

Yeah, that's the idea in STONE-459. PaC wouldn't really render the output into a UI element - it would render the output into a call to the GitHub Checks API, which would make a successful TaskRun with "errors" in its HACBS_TEST_OUTPUT end up being represented as a "failing" check in the GitHub UI.

> Instead of just having the UI/PaC go through all TaskRuns, would it be reasonable to have a "finalizer" Task which can aggregate and summarize the result of the previous scanning/linting TaskRuns.

This is possible, but PaC and the UI both already need to deal with per-task stuff anyways. PaC submits a check result for each task (in GitHub you can see a table of all ~dozen steps performed in a pipelinerun) and in the UI we render the whole pipeline run with the ability to click through into the logs of a task. Having both of those systems be able to render errors on a per-task level directs the user to exactly where they need to click to go see the stdout from scanner $foo.